### ISSUE TITLE
Stream segments of a FITS file over network protocols

### ISSUE THREAD

**Issue by keflavich** [astropy/astropy#3446]
This is mostly directed at @embray, I think.  

The [CFITSIO docs](http://heasarc.gsfc.nasa.gov/docs/software/fitsio/c/c_user/node82.html) indicate that FITS files can be opened read-only over a network protocol.  `astropy.io.fits` tries to download the file if its a URL, though.  Is it possible to enable network access to FITS files via pyfits?  There's a major use case in getting quicklook cutouts from large data cubes that I would like to exploit (especially in `spectral_cube`.

**Comment by Cadair** [comment]
@keflavich oh that would be cool.

**Comment by embray** [comment]
I want to do a total overhaul though of how pyfits handles I/O, and in particular I want to make it possible to read in, for example, just a single HDU without having to scan for every HDU in a file.  For your example of large data cubes over the network this would be especially important, as we would want to read only ranges of the file.  That level of support is definitely not possible at the moment (though possibly within the scope of the I/O rework).

**Comment by embray** [comment]
I changed the title of this since, as you noted, it's already possible to open a FITS file via HTTP.  The difference being that simply downloads the entire file, and does not stream it.

**Comment by keflavich** [comment]
See also https://gist.github.com/astrofrog/404a5682311bae221fd5308c1498da94 (thanks @astrofrog, @pllim)

**Comment by jbcurtin** [comment]
hello, I explored some of this work in `cloud-fits`. Please take the best parts and integrate it into this issue request
https://github.com/jbcurtin/cloud-fits

FITs extention headers do a great job predicting where data begins and ends. A parser could be written to scan the file for initial headers. Use the data to compute the location of the next header, until it finds all headers

The operations can further be optimized by deciding how to render the fits files

**Comment by Cadair** [comment]
Is this addressed by #13238 ?

**Comment by barentsen** [comment]
I should mention that #13238 does not enable cutouts to be obtained from compressed image extensions due to limitations in AstroPy's support for tile compression.  Other issues describe these limitations in more detail however (e.g., #3895), and any future improvements in AstroPy's support for tile compression should work seamlessly in the cloud because #13238 implemented generic cloud access at the file-object level.

TL;DR I think this issue can be closed!

**Comment by Cadair** [comment]
I came across this issue while looking for issues related to our compressed image work :wink:
